dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -164,32 +205,42 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
                     {
                         auto appUserModelId{ PushNotificationHelpers::GetAppUserModelId() };
                         THROW_IF_FAILED(CreateChannelWithRemoteIdHelper(appUserModelId, remoteId, channelInfo));
+
+                        // Need to recreate the Foreground sinks since channel creation removes the old sink
+                        RegisterSinkHelper();
                     }
                     else
                     {
+                        auto notificationPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+
                         // AppId is generated by PushNotificationLongRunningTask singleton
                         wil::unique_cotaskmem_string unpackagedAppUserModelId;
-                        RegisterUnpackagedApplicationHelper(remoteId, unpackagedAppUserModelId);
-                        THROW_IF_FAILED(CreateChannelWithRemoteIdHelper(unpackagedAppUserModelId, remoteId, channelInfo));
+                        THROW_IF_FAILED(notificationPlatform->RegisterFullTrustApplication(m_processName.get(), remoteId, &unpackagedAppUserModelId));
 
-                        auto notificationPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+                        THROW_IF_FAILED(CreateChannelWithRemoteIdHelper(unpackagedAppUserModelId, remoteId, channelInfo));
 
-                        wil::unique_cotaskmem_string processName;
-                        THROW_IF_FAILED(GetCurrentProcessPath(processName));
-                        THROW_IF_FAILED(notificationPlatform->RegisterLongRunningActivatorWithClsid(processName.get(), GetComServerClsid()));
+                        // Need to recreate the Foreground sinks since channel creation removes the old sink
+                        RegisterSinkHelper();
 
                         std::wstring toastAppId{ RetrieveNotificationAppId() };
-                        THROW_IF_FAILED(notificationPlatform->AddToastRegistrationMapping(processName.get(), toastAppId.c_str()));
+                        THROW_IF_FAILED(notificationPlatform->AddToastRegistrationMapping(m_processName.get(), toastAppId.c_str()));
+                    }
+
+                    auto channel{ winrt::make<PushNotificationChannel>(channelInfo) };
+                    {
+                        auto lock{ m_lock.lock_exclusive() };
+                        m_channel = channel;
                     }
 
                     PushNotificationTelemetry::ChannelRequestedByApi(S_OK, remoteId);
                     co_return winrt::make<PushNotificationCreateChannelResult>(
-                        winrt::make<PushNotificationChannel>(channelInfo),
+                        channel,
                         S_OK,
                         PushNotificationChannelStatus::CompletedSuccess);
                 }
                 catch (...)
                 {
+                    UnregisterSinkHelper(); // Need to clean up sinks if failure occurs.
```

Comment:
This should not be here. LEts not do this at all please

Reply:
It needs a scope exit and a dismiss pattern

Reply:
And probably outside the try catch. I would have a bool and use that to determine if channel request succeeded. If something failed, then cleanup after the catch

Reply:
@pmpurifoy